gax/status_or.h

```diff
@@ -0,0 +1,214 @@
+ * looks like usage of a smart pointer, or even a std::optional<T>, in that you
+ * first check its validity using a conversion to bool (or by calling
+ * `StatusOr::ok()`), then you may dereference the object to access the
+ * contained value. It is undefined behavior (UB) to dereference a
+ * `StatusOr<T>` that is not "ok". For example:
+ *
+ * @code
+ * StatusOr<Foo> foo = FetchFoo();
+ * if (!foo) {  // Same as !foo.ok()
+ *   // handle error and probably look at foo.status()
+ * } else {
+ *   foo->DoSomethingFooey();  // UB if !foo
+ * }
+ * @endcode
+ *
+ * Alternatively, you may call the `StatusOr::value()` member function.
+ *
+ * @code
+ * StatusOr<Foo> foo = FetchFoo();
+ * foo.value().DoSomethingFooey();  // May throw/crash if there is no value
+ * @endcode
+ *
+ * Functions that can fail will often return a `StatusOr<T>` instead of
+ * returning an error code and taking a `T` out-param, or rather than directly
+ * returning the `T` and throwing an exception on error. StatusOr is used so
+ * that callers can choose whether they want to explicitly check for errors,
+ * crash the program, or throw exceptions. Since constructors do not have a
+ * return value, they should be designed in such a way that they cannot fail by
+ * moving the object's complex initialization logic into a separate factory
+ * function that itself can return a `StatusOr<T>`. For example:
+ *
+ * @code
+ * class Bar {
+ *  public:
+ *   Bar(Arg arg);
+ *   ...
+ * };
+ * StatusOr<Bar> MakeBar() {
+ *   ... complicated logic that might fail
+ *   return Bar(std::move(arg));
+ * }
+ * @endcode
+ *
+ * TODO(...) - the current implementation is fairly naive with respect to `T`,
+ *   it is unlikely to work correctly for reference types, arrays, and so forth.
+ *
+ * @tparam T the type of the value.
+ */
+template<typename T>
+class StatusOr final {
+ public:
+  // There is no good definition of a default StatusOr:
+  // it can't default construct T because T may not have a default constructor,
+  // and using an unknown error code and generic message is not helpful.
+  StatusOr() = delete;
+
+  /**
+   * Creates a new `StatusOr<T>` holding the error condition @p rhs.
+   * It is undefined behavior to create a `StatusOr<T>` from an OK status.
+   *
+   * @par Post-conditions
+   * `ok() == false` and `status() == rhs`.
+   *
+   * @param rhs the status to initialize the object.
+   */
+  StatusOr(Status rhs) : status_(std::move(rhs)) {
+    if(status_.IsOk()){
+      std::cerr << "Constructing StatusOr<T> from OK status is not allowed" << std::endl;
+      std::abort();
+    }
+  }
+
+  /**
+   * Creates a new `StatusOr<T>` holding the value @p rhs.
+   *
+   * @par Post-conditions
+   * `ok() == true` and `value() == rhs`.
+   *
+   * @param rhs the value used to initialize the object.
+   */
+  StatusOr(T const& rhs) : status_(), value_(rhs) {}
+  StatusOr(T&& rhs) : status_(), value_(std::move(rhs)) {}
```

Comment:
Google Cloud C++ uses a call to placement new here. Is there a reason that we should or shouldn't do that?
https://github.com/googleapis/google-cloud-cpp/blob/master/google/cloud/status_or.h#L211

Reply:
The rule is that placement new _must_ be used when constructing in the function body, but we can call the constructor normally when the union member is part of an initializer list. We need to use placement new in the example below because we need more complicated logic to determine which member of the union to construct:
```
class Mollusc {
public:
Mollusc(bool isSquid) : isSquid_(isSquid) {
   if(isSquid_) {
      new (&squidName_) string("Architeuthis dux");
   } else {
      new (&clamName_) string("Tridacna squamosa");
   }
}

bool isSquid_;
union {
   string squidName_;
   string clamName_;
};
};
```
I have no strong feelings on the matter in the StatusOr constructors; the constructors that take a T, using the initializer list is more compact and clean, but we have to use placement new in the copy/move constructors, so maybe consistency is better. WDYT?

Reply:
I have no strong feelings either, which suggests to me that we should go with consistency + following the google-cloud-cpp implementation.

Reply:
Done